lib/ocpp/v201/device_model.cpp

```diff
@@ -367,5 +367,98 @@ void DeviceModel::check_integrity(const std::map<int32_t, int32_t>& evse_connect
     }
 }
 
+std::vector<SetMonitoringResult> DeviceModel::set_monitors(const std::vector<SetMonitoringData>& requests) {
+    if (requests.size() <= 0) {
+        return {};
+    }
+
+    std::vector<SetMonitoringResult> set_monitors_res;
+
+    for (auto& request : requests) {
+        SetMonitoringResult result;
+
+        if (this->device_model.find(request.component) == this->device_model.end()) {
+            result.status = SetMonitoringStatusEnum::UnknownComponent;
+            set_monitors_res.push_back(result);
+            continue;
+        }
+
+        auto variable_map = this->device_model[request.component];
+
+        if (variable_map.find(request.variable) == variable_map.end()) {
+            result.status = SetMonitoringStatusEnum::UnknownVariable;
+            set_monitors_res.push_back(result);
+            continue;
+        }
+
+        // TODO (ioan): see how we should handle the 'Duplicate' data
+        if (this->storage->set_monitoring_data(request)) {
+            result.status = SetMonitoringStatusEnum::Accepted;
+        } else {
+            result.status = SetMonitoringStatusEnum::Rejected;
+        }
+
+        set_monitors_res.push_back(result);
+    }
+
+    return set_monitors_res;
+}
+std::vector<MonitoringData> DeviceModel::get_monitors(const std::vector<MonitoringCriterionEnum>& criteria,
+                                                      const std::vector<ComponentVariable>& component_variables) {
+    if (criteria.size() <= 0 || component_variables.size() <= 0) {
+        return {};
+    }
+
+    std::vector<MonitoringData> get_monitors_res;
+
+    for (auto& component_variable : component_variables) {
+        if (component_variable.variable.has_value() == false) {
+            continue;
+        }
+
+        if (this->device_model.find(component_variable.component) == this->device_model.end()) {
+            continue;
+        }
+
+        auto variable_map = this->device_model[component_variable.component];
+
+        if (variable_map.find(component_variable.variable.value()) == variable_map.end()) {
+            continue;
+        }
+
+        // Search for all monitors related to the variable id
+        std::optional<MonitoringData> monitor_data = this->storage->get_monitoring_data(
```

Comment:
Question: should we have a fallback here? If we can't find a value in-memory should we try to retrieve it from the DB? However if we are doing that one time at startup it would not be required.